docs/architecture/ADR-001-ABCI++.md

```diff
@@ -0,0 +1,414 @@
+   l := len(txs)
+   bzs := make([][]byte, l)
+   for i := 0; i < l; i++ {
+       bzs[i] = txs[i]
+   }
+
+   preparedProposal, err := blockExec.proxyApp.PrepareProposalSync(
+       abci.RequestPrepareProposal{
+           BlockData:     &tmproto.Data{Txs: txs.ToSliceOfBytes(), Evidence: *pevdData},
+           BlockDataSize: maxDataBytes},
+   )
+   ...
+   return state.MakeBlock(
+       ...
+   )
+}
+```
+
+Here is where arbitrary logic is performed by the app before voting on a proposed block during consensus.
+
+```go
+func (cs *State) defaultDoPrevote(height int64, round int32) {
+  ...
+   stateMachineValidBlock, err := cs.blockExec.ProcessProposal(cs.ProposalBlock)
+   if err != nil {
+       cs.Logger.Error("state machine returned an error when trying to process proposal block", "err", err)
+   }
+
+   // Vote nil if application invalidated the block
+   if !stateMachineValidBlock {
+       // Consensus says we must vote nil
+       logger.Error("prevote step: consensus deems this block to be mustVoteNil", "err", err)
+       cs.signAddVote(tmproto.PrevoteType, nil, types.PartSetHeader{})
+       return
+   }
+
+   // Prevote cs.ProposalBlock
+   // NOTE: the proposal signature is validated when it is received,
+   // and the proposal block parts are validated as they are received (against the merkle hash in the proposal)
+   logger.Debug("prevote step: ProposalBlock is valid")
+   cs.signAddVote(tmproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+}
+```
+
+For those interested in how to incorporate these methods into the cosmos-sdk and the app, we do that in the following PRs.
+
+- celestiaorg/cosmos-sdk [#63](https://github.com/celestiaorg/cosmos-sdk/pull/63)
+- celestiaorg/celestia-app [#214](https://github.com/celestiaorg/celestia-app/pull/214)
+
+### PrepareProposal [#637](https://github.com/celestiaorg/celestia-core/pull/637) and [#224](https://github.com/celestiaorg/celestia-app/pull/224)
+
+The way that we create proposal blocks will be refactored (previously [`PrePreprocessTxs`](https://github.com/celestiaorg/celestia-app/blob/0363f0410d9d6bf0e51ac92afcaa9be7c0d1ba07/app/abci.go#L17-L108)) to accommodate the new features.
+
+```go
+// PrepareProposal separates messages from transactions, malleates those transactions,
+// estimates the square size, fills the data square, and calculates the data hash from
+// the erasure data
+func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
+   squareSize := app.estimateSquareSize(req.BlockData)
+
+   dataSquare, data, err := SplitShares(app.txConfig, squareSize, req.BlockData)
+   if err != nil {
+       panic(err)
+   }
+
+   eds, err := da.ExtendShares(squareSize, dataSquare)
+   if err != nil {
+       // ... log error
+       panic(err)
+   }
+
+   dah := da.NewDataAvailabilityHeader(eds)
+   data.Hash = dah.Hash() // <-- here we are setting the data hash before we pass the block data back to tendermint
```

Comment:
This means we can delete some more of the custom logic in tendermint or will still compute this in core additionally?

Reply:
Can we clarify what core still has to do still and what is taken over by the app now?

Reply:
yes, that's a great idea https://github.com/celestiaorg/celestia-app/pull/231/commits/40861a89e557e23c301e772ca8420440dbe366eb

as mentioned in the new paragraph, it does not allow us to remove that much code from tendermint, but it does dramatically increase efficiency as we are no long computing the hash every time we decode the block, and allows us to keep the rest of tendermint that expects this functionality the same (mainly tests).